### PR TITLE
Make getting a frozen partial snapshot async.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Followup to https://github.com/dotnet/roslyn/pull/71999.

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs [hunk 1]
```diff
@@ -1006,64 +1010,112 @@ public SolutionCompilationState WithOptions(SolutionOptionSet options)
...
-        if (this.Services.GetService<IWorkspacePartialSolutionsTestHook>()?.IsPartialSolutionDisabled == true)
-            return this;
+        AsyncLazy<SolutionCompilationState>? lazyState;
+        using (_cachedFrozenDocumentStateLock.DisposableWait(cancellationToken))
+        {
+            if (!_cachedFrozenDocumentState.TryGetValue(documentId, out lazyState))
+            {
+                lazyState = new AsyncLazy<SolutionCompilationState>(
+                    cancellationToken => ComputeFrozenPartialCompilationIncludingSpecificDocumentAsync(documentId, synchronous: false, cancellationToken),
+                    cancellationToken => ComputeFrozenPartialCompilationIncludingSpecificDocumentAsync(documentId, synchronous: true, cancellationToken).GetAwaiter().GetResult());
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs:1064]: >cancellationToken => ComputeFrozenPartialCompilationIncludingSpecificDocumentAsync(documentId, synchronous: true, cancellationToken).GetAwaiter().GetResult() [](http://example.com/codeflow?start=20&length=157)

Learning time: What's the advantage of doing this over not sending in a synchronous function?

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs]: updated this.  it's a but more code duplication, but makes it clear how one function is sync, and the other async.  teh benefit of hte original pattern is sharing as much code as possible.  However, it is a big uglier as you have to be very careful to do no async work when `synchronous: true` is passed in, and you then have to know it's not actually going to block the returned Task when .GetAwaiter().Result is used.

**Comment by CyrusNajmabadi** [comment]
Goign to possibly table this.  @jasonmalinowski  makes a good poitn that we want to keep this sync as we want to use this in codepaths that want to try to stay sync/fast as much as possible.  becoming async makes it very easy for this to start exhibiting bad async behavior.

**Comment by jasonmalinowski** [comment]
@CyrusNajmabadi I did have one thought over the weekend -- if the remaining "expensive" bit of that path is the parse, we can probably make a change where we instead push that parsing over to the later call for Document.GetSyntaxTreeAsync/Project.GetCompilationAsync(); I think how it would have to work though is rather than the CompilationTracker getting a FinalState one, we'd have to create a InProgressState with a special translation action that did the equivalent of

https://github.com/dotnet/roslyn/blob/a3c0cbd96fa27cded000dacb8114a7ab3d475420/src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs#L204-L259

It would probably be a smaller change, and would mean that if you have a bunch of features asking for the frozen thing we'd be able to get that new Solution snapshot out as fast as possible, and then we'd be able to do stuff from there like normal.

**Comment by CyrusNajmabadi** [comment]
> can probably make a change where we instead push that parsing over to the later call for Document.GetSyntaxTreeAsync/Project.GetCompilationAsync(); I think how it would have to work though is rather than the CompilationTracker getting a FinalState one, we'd have to create a InProgressState with a special translation action that did the equivalent of

I really like that idea.  I'll look into that :)

**Pull request closed without merging**